edalize/libero.py

```diff
@@ -0,0 +1,181 @@
+
+class Libero(Edatool):
+    @classmethod
+    def get_doc(cls, api_ver):
+        if api_ver == 0:
+            return {'description': "The Libero backend supports Microsemi Libero to build systems and program the FPGA",
+                    'members': [
+                        {'name': 'family',
+                         'type': 'String',
+                         'desc': 'FPGA family (e.g. PolarFire)'},
+                        {'name': 'die',
+                         'type': 'String',
+                         'desc': 'FPGA device (e.g. MPF300TS)'},
+                        {'name': 'package',
+                         'type': 'String',
+                         'desc': 'FPGA package type (e.g. FCG1152)'},
+                        {'name': 'speed',
+                         'type': 'String',
+                         'desc': 'FPGA speed rating (e.g. -1)'},
+                        {'name': 'dievoltage',
+                         'type': 'String',
+                         'desc': 'FPGA die voltage (e.g. 1.0)'},
+                        {'name': 'range',
+                         'type': 'String',
+                         'desc': 'FPGA temperature range (e.g. IND)'},
+                        {'name': 'defiostd',
+                         'type': 'String',
+                         'desc': 'FPGA default IO std (e.g. "LVCMOS 1.8V"'},
+                        {'name': 'hdl',
+                         'type': 'String',
+                         'desc': 'Default HDL (e.g. "VERILOG"'},
+                        {'name': 'synth_options',
+                         'type': 'String',
+                         'desc': 'Additional Synthesize tool options separated by commas (e.g. "RETIMING:true,CLOCK_GLOBAL:2")'},
+                        {'name': 'pnr_options',
+                         'type': 'String',
+                         'desc': 'Additional Place and Route tool options separated by commas (e.g. "REPAIR_MIN_DELAY:true,TDPR:true")'},
+                    ]
+                    }
+
+    argtypes = ['vlogdefine', 'vlogparam']
+    mandatory_options = ['family', 'die', 'package', 'range']
+
+    tool_options_defaults = {
+        'range': 'IND',
+    }
+
+    def _get_tool_params(self, params):
+        d = dict()
+        for p in params.split(","):
+            d[p.split(":")[0]] = p.split(":")[1]
+        return d
+
+    def _set_tool_options_defaults(self):
+        for key, default_value in self.tool_options_defaults.items():
+            if not key in self.tool_options:
+                logger.info("Set Libero tool option %s to default value %s"
+                            % (key, str(default_value)))
+                self.tool_options[key] = default_value
+
+    def _check_mandatory_options(self):
+        shouldExit = 0
+        for key in self.mandatory_options:
+            if not key in self.tool_options:
+                logger.error("Libero option \"%s\" must be defined", key)
+                shouldExit = 1
+        if shouldExit:
```

Comment:
We can't use exit in the backends. Raise RuntimeError instead

Reply:
Done!